crates/rethnet_eth/src/remote/cacheable_method_invocation.rs

```diff
@@ -0,0 +1,680 @@
+            address,
+        } = params;
+
+        let this = self
+            .hash_block_spec(from_block)?
+            .hash_block_spec(to_block)?
+            .hash_bytes(address.as_bytes());
+        Some(this)
+    }
+
+    fn hash_method_invocation(self, method: &CacheableMethodInvocation<'_>) -> Option<Self> {
+        use CacheableMethodInvocation::*;
+
+        let this = self.hash_u8(method.cache_key_variant());
+
+        let this = match method {
+            ChainId => this,
+            GetBalance {
+                address,
+                block_spec,
+            } => this
+                .hash_address(address)
+                .hash_maybe_block_spec(block_spec)?,
+            GetBlockByNumber {
+                block_spec,
+                include_tx_data,
+            } => this.hash_block_spec(block_spec)?.hash_bool(include_tx_data),
+            GetBlockByHash {
+                block_hash,
+                include_tx_data,
+            } => this.hash_b256(block_hash).hash_bool(include_tx_data),
+            GetBlockTransactionCountByHash { block_hash } => this.hash_b256(block_hash),
+            GetBlockTransactionCountByNumber { block_spec } => this.hash_block_spec(block_spec)?,
+            GetCode {
+                address,
+                block_spec,
+            } => this
+                .hash_address(address)
+                .hash_maybe_block_spec(block_spec)?,
+            GetLogs { params } => this.hash_get_logs_input(params)?,
+            GetStorageAt {
+                address,
+                position,
+                block_spec,
+            } => this
+                .hash_address(address)
+                .hash_u256(position)
+                .hash_maybe_block_spec(block_spec)?,
+            GetTransactionByBlockHashAndIndex { block_hash, index } => {
+                this.hash_b256(block_hash).hash_u256(index)
+            }
+            GetTransactionByBlockNumberAndIndex {
+                block_number,
+                index,
+            } => this.hash_u256(block_number).hash_u256(index),
+            GetTransactionByHash { transaction_hash } => this.hash_b256(transaction_hash),
+            GetTransactionCount {
+                address,
+                block_spec,
+            } => this
+                .hash_address(address)
+                .hash_maybe_block_spec(block_spec)?,
+            GetTransactionReceipt { transaction_hash } => this.hash_b256(transaction_hash),
+            NetVersion => this,
+        };
+
+        Some(this)
+    }
+
+    fn hash_method_invocations(self, methods: &CacheableMethodInvocations<'_>) -> Option<Self> {
+        // Make sure it's prefix-free
+        let mut this = self.hash_usize(methods.len());
+
+        for method_invocation in methods {
+            this = this.hash_method_invocation(method_invocation)?;
+        }
+
+        Some(this)
+    }
+
+    fn finalize(self) -> Option<CacheKey> {
+        // Returns Option for chaining convenience.
```

Comment:
Thanks good catch! Fixed in [0b44369](https://github.com/NomicFoundation/hardhat/pull/4294/commits/0b44369b890ae09c519bd4283c635aa2122b4ef4). (The comment and the `Option` return type there are a left over from when the code had different structure and otherwise would've had to wrap `finalize` calls in `Some` all the time in this module.)

In other places like [this](https://github.com/NomicFoundation/hardhat/pull/4294/files#diff-cf941b21bd00797395a8590c975818931f27bb21a221df240c931cfd0b317a9cR402) where the return value is `Some(result)` we are using short-circuiting with `?`, so those methods can return `None`.